core/src/main/java/org/apache/brooklyn/core/OsgiLauncher.java

```diff
@@ -0,0 +1,557 @@
+            if (!catInit.hasRunOfficialInitialization()) {
+                if (persistMode == PersistMode.DISABLED) {
+                    log.debug("Loading catalog as part of launch sequence (it was not loaded as part of any rebind sequence)");
+                    catInit.populateCatalog(ManagementNodeState.MASTER, true, true, null);
+                } else {
+                    // should have loaded during rebind
+                    ManagementNodeState state = managementContext.getHighAvailabilityManager().getNodeState();
+                    log.warn("Loading catalog for " + state + " as part of launch sequence (it was not loaded as part of the rebind sequence)");
+                    catInit.populateCatalog(state, true, true, null);
+                }
+            }
+        } catch (Exception e) {
+            handleSubsystemStartupError(ignoreCatalogErrors, "initial catalog", e);
+        }
+        catInit.setStartingUp(false);
+
+        // no need for command-line locations
+//        // Create the locations. Must happen after persistence is started in case the
+//        // management context's catalog is loaded from persisted state. (Location
+//        // resolution uses the catalog's classpath to scan for resolvers.)
+//        locations.addAll(managementContext.getLocationRegistry().resolve(locationSpecs));
+        // Already rebinded successfully, so previous apps are now available.
+        // Allow the startup to be visible in console for newly created apps.
+        ((LocalManagementContext) managementContext).noteStartupComplete();
+
+//        // TODO create apps only after becoming master, analogously to catalog initialization
+//        try {
+//            createApps();
+//            startApps();
+//        } catch (Exception e) {
+//            handleSubsystemStartupError(ignoreAppErrors, "brooklyn autostart apps", e);
+//        }
+//        if (startBrooklynNode) {
+//            try {
+//                startBrooklynNode();
+//            } catch (Exception e) {
+//                handleSubsystemStartupError(ignoreAppErrors, "brooklyn node / self entity", e);
+//            }
+//        }
+//
+        if (persistMode != PersistMode.DISABLED) {
+            // Make sure the new apps are persisted in case process exits immediately.
+            managementContext.getRebindManager().forcePersistNow(false, null);
+        }
+    }
+
+    public void shutdown() {
+        // not needed: webserver is in other bundles
+//        if (webServer != null) {
+//            try {
+//                webServer.stop();
+//            } catch (Exception e) {
+//                LOG.warn("Error stopping web-server; continuing with termination", e);
+//            }
+//        }
+
+        // TODO Do we want to do this as part of managementContext.terminate, so after other threads are terminated etc?
+        // Otherwise the app can change between this persist and the terminate.
```

Comment:
Answered in a comment above. The first thing to define is what "shutdown" means in the OSGi context. My opinion is not to consider `brooklyn-core` bundle deactivation as shutdown, since that happens when we want to upgrade it, or might happen when manipulating karaf cellar nodes.